File: r/pragmastat/tests/testthat/test-assume-sorted.R
# Tests for the raw (vector) API's `assume_sorted` flag.
#
# The vector path is otherwise only exercised with assume_sorted = FALSE; the
# assume_sorted = TRUE branch is reached transitively via Sample. These tests hit
# it directly.
#
# For ORDER-INDEPENDENT estimators (center, spread, shift, ratio, disparity, and
# the center/shift/ratio bounds), calling with assume_sorted = TRUE on a sorted
# copy must equal calling with assume_sorted = FALSE on the unsorted input.
#
# For SHUFFLE-based bounds (spread_bounds, disparity_bounds), the flag only
# affects the sparity check and never the shuffle, so for the SAME input and
# SAME seed the result must be identical regardless of the flag. We compare on
# SORTED input: the shuffle runs on the passed order (identical in both calls),
# and sorted input keeps the order-independent sparity check well-defined.
# Passing assume_sorted = TRUE on UNSORTED input is documented undefined
# behavior (the C sparity kernel relies on the sorted invariant), so we do not
# exercise that misuse here.

tol <- 1e-9

x_unsorted <- c(5, 2, 8, 1, 9, 3, 7, 4)
y_unsorted <- c(6, 2, 9, 1, 4, 7, 3, 8)
x_sorted <- sort(x_unsorted)
y_sorted <- sort(y_unsorted)

test_that("center assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    center(x_sorted, assume_sorted = TRUE),
    center(x_unsorted, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("spread assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    spread(x_sorted, assume_sorted = TRUE),
    spread(x_unsorted, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("shift assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    shift(x_sorted, y_sorted, assume_sorted = TRUE),
    shift(x_unsorted, y_unsorted, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("ratio assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    ratio(x_sorted, y_sorted, assume_sorted = TRUE),
    ratio(x_unsorted, y_unsorted, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("disparity assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    disparity(x_sorted, y_sorted, assume_sorted = TRUE),
    disparity(x_unsorted, y_unsorted, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("center_bounds assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    center_bounds(x_sorted, misrate = 0.3, assume_sorted = TRUE),
    center_bounds(x_unsorted, misrate = 0.3, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("shift_bounds assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    shift_bounds(x_sorted, y_sorted, misrate = 0.3, assume_sorted = TRUE),
    shift_bounds(x_unsorted, y_unsorted, misrate = 0.3, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("ratio_bounds assume_sorted=TRUE on sorted input equals unsorted call", {
  expect_equal(
    ratio_bounds(x_sorted, y_sorted, misrate = 0.3, assume_sorted = TRUE),
    ratio_bounds(x_unsorted, y_unsorted, misrate = 0.3, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("spread_bounds is flag-invariant on the same input and seed", {
  # spread_bounds shuffles internally; the assume_sorted flag only affects the
  # sparity check, never the shuffle. Same input + same seed must be identical.
  expect_equal(
    spread_bounds(x_sorted, misrate = 0.3, seed = 42, assume_sorted = TRUE),
    spread_bounds(x_sorted, misrate = 0.3, seed = 42, assume_sorted = FALSE),
    tolerance = tol
  )
})

test_that("disparity_bounds is flag-invariant on the same input and seed", {
  expect_equal(
    disparity_bounds(x_sorted, y_sorted, misrate = 0.3, seed = 42, assume_sorted = TRUE),
    disparity_bounds(x_sorted, y_sorted, misrate = 0.3, seed = 42, assume_sorted = FALSE),
    tolerance = tol
  )
})

# NOTE on the spread_bounds UNSORTED-inertness property
# -----------------------------------------------------
# There is deliberately NO "spread_bounds inert on UNSORTED input" test (i.e.
#   spread_bounds(UNSORTED, ..., assume_sorted = TRUE) ==
#   spread_bounds(UNSORTED, ..., assume_sorted = FALSE) byte-for-byte).
#
# With assume_sorted = TRUE the sparity check feeds the unsorted vector to the
# sorted-only C kernel spread_impl_c(arr, assume_sorted = 1), which is
# UNDEFINED BEHAVIOR: the selection loop cannot make monotone progress and
# bails out via the iteration cap + stall guard with a deterministic
# "Convergence failure (pathological input)" error (see the pathological-input
# regression test below), while the FALSE call sorts a copy and succeeds — on
# UNSORTED input the two flag values do not agree, so the equality would not
# hold. This matches the other language implementations, which also reject
# this assertion. The legitimate path always feeds a truly-sorted view
# (Sample$sorted_values, or a native vector the caller actually sorted), under
# which the "spread_bounds is flag-invariant on the same input and seed" test
# above already provides the coverage.

# Mutation / aliasing guard. The C kernels sort a copy internally (or, with
# assume_sorted = TRUE on legitimately sorted input, read the buffer strictly
# read-only). Every raw estimator must leave the caller's vector byte-for-byte
# unchanged under BOTH flag values, and a Sample's cached $sorted_values must
# survive estimator calls that alias it read-only.
test_that("raw estimators do not mutate the caller's vector (assume_sorted FALSE/TRUE)", {
  m <- 0.3

  for (flag in c(FALSE, TRUE)) {
    # With assume_sorted = TRUE feed the genuinely sorted vectors (the contract);
    # with FALSE feed the unsorted vectors.
    xv <- if (flag) x_sorted else x_unsorted
    yv <- if (flag) y_sorted else y_unsorted

    # Snapshots must be GENUINE COPIES (`+ 0` forces a fresh allocation).
    # A plain `x_before <- xv` binds the same underlying buffer, so a C kernel
    # writing through REAL() would mutate both bindings at once and the
    # expect_identical() below could never fail.

    # One-sample point + bounds estimators.
    one_sample <- list(
      function(v) center(v, assume_sorted = flag),
      function(v) spread(v, assume_sorted = flag),
      function(v) center_bounds(v, misrate = m, assume_sorted = flag),
      function(v) spread_bounds(v, misrate = m, seed = 42, assume_sorted = flag)
    )
    for (est in one_sample) {
      x_before <- xv + 0
      est(xv)
      expect_identical(xv, x_before)
    }

    # Two-sample point + bounds estimators.
    two_sample <- list(
      function(a, b) shift(a, b, assume_sorted = flag),
      function(a, b) ratio(a, b, assume_sorted = flag),
      function(a, b) disparity(a, b, assume_sorted = flag),
      function(a, b) shift_bounds(a, b, misrate = m, assume_sorted = flag),
      function(a, b) ratio_bounds(a, b, misrate = m, assume_sorted = flag),
      function(a, b) disparity_bounds(a, b, misrate = m, seed = 42, assume_sorted = flag)
    )
    for (est in two_sample) {
      x_before <- xv + 0
      y_before <- yv + 0
      est(xv, yv)
      expect_identical(xv, x_before)
      expect_identical(yv, y_before)
    }
  }
})

# center() on UNSORTED input with assume_sorted = TRUE is documented undefined
# behavior, but it must NEVER wedge the process in an unkillable infinite loop
# inside the C extension. The center_impl selection loop has an iteration cap +
# stall guard that raises a deterministic convergence error (a plain error, NOT
# an assumption_error) on pathological input. This test asserts the call fails
# fast rather than hanging.
test_that("center on unsorted input with assume_sorted=TRUE fails fast, never hangs", {
  # A deliberately unsorted vector chosen to violate the partition invariant
  # that Monahan's algorithm relies on; with assume_sorted = TRUE the loop
  # cannot make monotone progress and must bail out quickly.
  pathological <- c(100, 1, 99, 2, 98, 3, 97, 4, 96, 5)

  elapsed <- system.time(
    err <- tryCatch(
      {
        center(pathological, assume_sorted = TRUE)
        NULL
      },
      error = function(e) e
    )
  )["elapsed"]

  expect_true(!is.null(err), info = "expected a convergence error, got a value")
  expect_match(conditionMessage(err), "Convergence failure \\(pathological input\\)")
  # Must be a plain error, not an assumption_error.
  expect_false(inherits(err, "assumption_error"))
  # Fails fast: well under a second on any machine.
  expect_lt(as.numeric(elapsed), 5)
})

# spread() on UNSORTED input with assume_sorted = TRUE is documented undefined
# behavior, but it must NEVER wedge the process in an unkillable infinite loop
# inside the C extension. The spread_impl selection loop mirrors center_impl's
# iteration cap + stall guard and raises a deterministic convergence error (a
# plain error, NOT an assumption_error) on pathological input. This test
# asserts the call fails fast rather than hanging.
test_that("spread on unsorted input with assume_sorted=TRUE fails fast, never hangs", {
  # A deliberately unsorted vector chosen to violate the sorted invariant that
  # the Shamos selection loop relies on; with assume_sorted = TRUE the loop
  # cannot make monotone progress and must bail out quickly.
  pathological <- c(100, 1, 99, 2, 98, 3, 97, 4, 96, 5)

  elapsed <- system.time(
    err <- tryCatch(
      {
        spread(pathological, assume_sorted = TRUE)
        NULL
      },
      error = function(e) e
    )
  )["elapsed"]

  expect_true(!is.null(err), info = "expected a convergence error, got a value")
  expect_match(conditionMessage(err), "Convergence failure \\(pathological input\\)")
  # Must be a plain error, not an assumption_error.
  expect_false(inherits(err, "assumption_error"))
  # Fails fast: well under a second on any machine.
  expect_lt(as.numeric(elapsed), 5)
})

test_that("Sample $sorted_values is unchanged after estimator calls (read-only aliasing)", {
  sx <- Sample$new(x_unsorted)
  sy <- Sample$new(y_unsorted)

  # Genuine copies (`+ 0`): the active binding hands out the cached buffer
  # itself, so a plain assignment would alias it and an in-place mutation by a
  # C kernel could never be detected by expect_identical().
  sorted_x_before <- sx$sorted_values + 0
  sorted_y_before <- sy$sorted_values + 0

  # First round of calls aliases the cached sorted buffers read-only.
  r1_center <- center(sx)$value
  r1_spread <- spread(sx)$value
  r1_shift <- shift(sx, sy)$value
  r1_ratio <- ratio(sx, sy)$value
  r1_disparity <- disparity(sx, sy)$value

  expect_identical(sx$sorted_values, sorted_x_before)
  expect_identical(sy$sorted_values, sorted_y_before)

  # Two successive calls must return identical results: the cached buffer was not
  # disturbed by the first call (pins the read-only aliasing invariant).
  expect_identical(center(sx)$value, r1_center)
  expect_identical(spread(sx)$value, r1_spread)
  expect_identical(shift(sx, sy)$value, r1_shift)
  expect_identical(ratio(sx, sy)$value, r1_ratio)
  expect_identical(disparity(sx, sy)$value, r1_disparity)

  expect_identical(sx$sorted_values, sorted_x_before)
  expect_identical(sy$sorted_values, sorted_y_before)
})

File: py/tests/test_assume_sorted.py
"""Direct coverage of the raw API's ``assume_sorted=True`` branch.

The reference suite exercises the raw estimators only with ``assume_sorted=False``;
the ``=True`` branch is otherwise reached only transitively via :class:`Sample`.
These tests hit it directly through the native-array API.

Two contracts are checked:

* ORDER-INDEPENDENT estimators (center, spread, shift, ratio, disparity,
  center_bounds, shift_bounds, ratio_bounds): sorting the input ascending and
  calling with ``assume_sorted=True`` must equal the call on the UNSORTED input
  with ``assume_sorted=False`` (within ~1e-9).
* SHUFFLE-based bounds (spread_bounds, disparity_bounds): ``assume_sorted`` only
  affects the order-independent sparity check, never the disjoint-pair shuffle,
  so on the SAME array and SAME seed the result must be IDENTICAL for
  ``assume_sorted=True`` vs ``False``. We use an already-SORTED array here so that
  the sparity check receives valid input under BOTH flag values (the kernels treat
  ``assume_sorted=True`` on unsorted input as undefined behavior); the shuffle then
  runs on the identical order regardless, so the two results must be byte-identical.
"""

import numpy as np
import pytest

import pragmastat.center_impl as center_impl_module
import pragmastat.spread_impl as spread_impl_module
from pragmastat import (
    center,
    center_bounds,
    disparity,
    disparity_bounds,
    ratio,
    ratio_bounds,
    shift,
    shift_bounds,
    spread,
    spread_bounds,
)

TOL = 1e-9


@pytest.fixture(params=["c", "python"])
def kernel(request, monkeypatch):
    """Route a test through a specific selection kernel (C extension or pure Python).

    ``_center_impl`` / ``_spread_impl`` dispatch on their module-level
    ``_HAS_C_EXTENSION`` flags at call time, so forcing the flags to ``False``
    sends the public estimators through ``_center_impl_python`` /
    ``_spread_impl_python``. Without this, a build with the C extension present
    would never exercise the pure-Python guard paths.
    """
    if request.param == "python":
        monkeypatch.setattr(center_impl_module, "_HAS_C_EXTENSION", False)
        monkeypatch.setattr(spread_impl_module, "_HAS_C_EXTENSION", False)
    elif not (center_impl_module._HAS_C_EXTENSION and spread_impl_module._HAS_C_EXTENSION):  # noqa: SLF001
        pytest.skip("C extension not built")
    return request.param


# Unsorted, strictly positive (so ratio is defined), 8 elements (large enough for
# the 0.3 misrate used by the bounds estimators).
X_UNSORTED = [5.0, 1.0, 8.0, 3.0, 2.0, 7.0, 4.0, 6.0]
Y_UNSORTED = [12.0, 9.0, 15.0, 10.0, 13.0, 11.0, 16.0, 14.0]
MISRATE = 0.3
SEED = "assume-sorted-roundtrip"


def _sorted(values):
    return np.sort(np.asarray(values, dtype=np.float64))


class TestOrderIndependentRoundtrip:
    """sorted(input)+assume_sorted=True == unsorted+assume_sorted=False."""

    def test_center(self):
        unsorted = center(X_UNSORTED, assume_sorted=False)
        presorted = center(_sorted(X_UNSORTED), assume_sorted=True)
        assert abs(presorted - unsorted) < TOL

    def test_spread(self):
        unsorted = spread(X_UNSORTED, assume_sorted=False)
        presorted = spread(_sorted(X_UNSORTED), assume_sorted=True)
        assert abs(presorted - unsorted) < TOL

    def test_shift(self):
        unsorted = shift(X_UNSORTED, Y_UNSORTED, assume_sorted=False)
        presorted = shift(_sorted(X_UNSORTED), _sorted(Y_UNSORTED), assume_sorted=True)
        assert abs(presorted - unsorted) < TOL

    def test_ratio(self):
        unsorted = ratio(X_UNSORTED, Y_UNSORTED, assume_sorted=False)
        presorted = ratio(_sorted(X_UNSORTED), _sorted(Y_UNSORTED), assume_sorted=True)
        assert abs(presorted - unsorted) < TOL

    def test_disparity(self):
        unsorted = disparity(X_UNSORTED, Y_UNSORTED, assume_sorted=False)
        presorted = disparity(_sorted(X_UNSORTED), _sorted(Y_UNSORTED), assume_sorted=True)
        assert abs(presorted - unsorted) < TOL

    def test_center_bounds(self):
        unsorted = center_bounds(X_UNSORTED, MISRATE, assume_sorted=False)
        presorted = center_bounds(_sorted(X_UNSORTED), MISRATE, assume_sorted=True)
        assert abs(presorted.lower - unsorted.lower) < TOL
        assert abs(presorted.upper - unsorted.upper) < TOL

    def test_shift_bounds(self):
        unsorted = shift_bounds(X_UNSORTED, Y_UNSORTED, MISRATE, assume_sorted=False)
        presorted = shift_bounds(_sorted(X_UNSORTED), _sorted(Y_UNSORTED), MISRATE, assume_sorted=True)
        assert abs(presorted.lower - unsorted.lower) < TOL
        assert abs(presorted.upper - unsorted.upper) < TOL

    def test_ratio_bounds(self):
        unsorted = ratio_bounds(X_UNSORTED, Y_UNSORTED, MISRATE, assume_sorted=False)
        presorted = ratio_bounds(_sorted(X_UNSORTED), _sorted(Y_UNSORTED), MISRATE, assume_sorted=True)
        assert abs(presorted.lower - unsorted.lower) < TOL
        assert abs(presorted.upper - unsorted.upper) < TOL


class TestShuffleBoundsFlagInvariant:
    """assume_sorted never changes shuffle-based bounds (same array, same seed).

    A SORTED array is used so the order-independent sparity check gets valid input
    under both flag values; the disjoint-pair shuffle runs on the identical order
    regardless of the flag, so the two results must be byte-identical.
    """

    X_SORTED = sorted(X_UNSORTED)
    Y_SORTED = sorted(Y_UNSORTED)

    def test_spread_bounds(self):
        false_result = spread_bounds(self.X_SORTED, MISRATE, SEED, assume_sorted=False)
        true_result = spread_bounds(self.X_SORTED, MISRATE, SEED, assume_sorted=True)
        assert true_result.lower == false_result.lower
        assert true_result.upper == false_result.upper

    def test_disparity_bounds(self):
        false_result = disparity_bounds(self.X_SORTED, self.Y_SORTED, MISRATE, SEED, assume_sorted=False)
        true_result = disparity_bounds(self.X_SORTED, self.Y_SORTED, MISRATE, SEED, assume_sorted=True)
        assert true_result.lower == false_result.lower
        assert true_result.upper == false_result.upper


class TestAssumeSortedMisuseTerminates:
    """``assume_sorted=True`` on UNSORTED input must terminate, not wedge.

    The Monahan selection loop in ``center``'s kernel relies on genuine
    sortedness. Handing it UNSORTED data with ``assume_sorted=True`` is undefined
    behavior; without an iteration cap the loop can run forever, wedging the
    process (an unkillable infinite loop inside the C extension that Python-level
    signals cannot interrupt). All kernels bound the loop (256 + 4*n iteration
    cap plus a stall guard on the active set) and raise a deterministic
    convergence error instead. We assert that the call returns quickly with that
    error rather than hanging, via the ``kernel`` fixture for BOTH the C
    extension and the pure-Python fallback.
    """

    # A pathological UNSORTED arrangement that drives the selection loop without
    # progress when (incorrectly) treated as sorted.
    UNSORTED = [3.0, 1.0, 4.0, 1.5, 5.0, 9.0, 2.0, 6.0, 0.5, 8.0, 7.0]

    @pytest.mark.usefixtures("kernel")
    def test_center_unsorted_assume_sorted_raises_quickly(self):
        import time

        start = time.monotonic()
        with pytest.raises(RuntimeError, match=r"[Cc]onvergence failure"):
            center(np.asarray(self.UNSORTED, dtype=np.float64), assume_sorted=True)
        # Must be near-instant; a hang would blow far past this budget.
        assert time.monotonic() - start < 5.0

    # An UNSORTED arrangement that drives spread_impl's selection loop without
    # progress when (incorrectly) treated as sorted. Verified to loop forever in
    # an UNCAPPED spread kernel (an unkillable hang), and to raise the
    # deterministic convergence error with the cap in place. (The center UNSORTED
    # array above does NOT exercise spread's failure mode, hence a dedicated
    # input here.)
    UNSORTED_SPREAD = [5.16, -1.59, -4.82, 0.23, -1.9, 5.68, -3.93, -0.47, 1.67]

    @pytest.mark.usefixtures("kernel")
    def test_spread_unsorted_assume_sorted_raises_quickly(self):
        import time

        start = time.monotonic()
        with pytest.raises(RuntimeError, match=r"[Cc]onvergence failure"):
            spread(np.asarray(self.UNSORTED_SPREAD, dtype=np.float64), assume_sorted=True)
        # Must be near-instant; a hang would blow far past this budget.
        assert time.monotonic() - start < 5.0


# NOTE on the disparity_bounds / spread_bounds UNSORTED behavior
# -------------------------------------------------------------
# There is deliberately NO test asserting that the shuffle-based bounds are
# inert under ``assume_sorted`` on UNSORTED input, because that inertness does
# NOT hold for either estimator:
#
# * ``spread_bounds``: the sparity check skips its re-sort and feeds the passed
#   (unsorted) buffer to the sorted-only spread kernel, so the flag changes the
#   computation path (``test_spread_unsorted_assume_sorted_raises_quickly``
#   shows the kernel can error on such input).
# * ``disparity_bounds``: its embedded order-independent shift-bounds consumes
#   the passed slice as a sorted view, so the flag can change the result.
#
# Passing ``assume_sorted=True`` on UNSORTED input is a contract violation
# (undefined behavior): the RESULT is unspecified, but TERMINATION is
# guaranteed. Without an iteration cap, both ``center``'s Monahan selection AND
# ``spread``'s selection loop can HANG on such input; the pure-Python and C
# extension paths all bound the loop (256 + 4*n iteration cap plus a stall
# guard on the active set) and raise a deterministic convergence error (see
# ``TestAssumeSortedMisuseTerminates`` above). The legitimate path always
# feeds a truly-sorted view (``Sample.sorted_values``, or a native array the
# caller actually sorted), under which ``TestShuffleBoundsFlagInvariant`` covers
# flag-invariance.
